src/QuickCheckVEngine/RVFI_DII/RVFI.hs

```diff
@@ -73,175 +84,404 @@ type RV_RegIdx = Word8
+
+rvfiDecodeV1Response :: Get RVFI_Packet
+rvfiDecodeV1Response = do
+  intr <- getWord8
+  halt <- getWord8
+  trap <- getWord8
+  rd_addr <- getWord8
+  rs2_addr <- getWord8
+  rs1_addr <- getWord8
+  mem_wmask <- getWord8
+  mem_rmask <- getWord8
+  mem_wdata <- getWord64be
+  mem_rdata <- getWord64be
+  mem_addr <- getWord64be
+  rd_wdata <- getWord64be
+  rs2_rdata <- getWord64be
+  rs1_rdata <- getWord64be
+  insn <- getWord64be
+  pc_wdata <- getWord64be
+  pc_rdata <- getWord64be
+  order <- getWord64be
+  return $
+    RVFI_Packet
+      { rvfi_valid = 1,
+        rvfi_order = order,
+        rvfi_insn = insn,
+        rvfi_trap = trap,
+        rvfi_halt = halt,
+        rvfi_intr = intr,
+        rvfi_mode = Nothing,
+        rvfi_ixl = Nothing,
+        rvfi_pc_rdata = pc_rdata,
+        rvfi_pc_wdata = pc_wdata,
+        rvfi_int_data =
+          Just
+            RVFI_IntData
+              { rvfi_rs1_addr = rs1_addr,
+                rvfi_rs2_addr = rs2_addr,
+                rvfi_rs1_rdata = rs1_rdata,
+                rvfi_rs2_rdata = rs2_rdata,
+                rvfi_rd_addr = rd_addr,
+                rvfi_rd_wdata = rd_wdata
+              },
+        rvfi_mem_data =
+          Just
+            RVFI_MemAccessData
+              { rvfi_mem_addr = mem_addr,
+                rvfi_mem_rmask = fromIntegral mem_rmask, -- FIXME: I think this zero-extends?
+                rvfi_mem_wmask = fromIntegral mem_wmask, -- FIXME: I think this zero-extends?
+                rvfi_mem_rdata = Basement.Types.Word256.Word256 0 0 0 mem_rdata,
+                rvfi_mem_wdata = Basement.Types.Word256.Word256 0 0 0 mem_wdata
+              }
+      }
 
 -- | An otherwise empty halt token for padding
 rvfiEmptyHaltPacket :: RVFI_Packet
-rvfiEmptyHaltPacket = (runGet get (BS.repeat 0)) { rvfi_halt = 1 }
+rvfiEmptyHaltPacket =
+  RVFI_Packet
+    { rvfi_halt = 1,
+      rvfi_valid = 0,
+      rvfi_order = 0,
+      rvfi_insn = 0,
+      rvfi_trap = 0,
+      rvfi_intr = 0,
+      rvfi_mode = Nothing,
+      rvfi_ixl = Nothing,
+      rvfi_pc_rdata = 0,
+      rvfi_pc_wdata = 0,
+      rvfi_int_data = Nothing,
+      rvfi_mem_data = Nothing
+    }
+
+prvString :: Maybe RV_PrivMode -> String
+prvString Nothing = "PRV_?"
+prvString (Just x) = show x
 
 instance Show RVFI_Packet where
   show tok
-    | rvfiIsHalt tok = "halt token"
-    | otherwise = printf "Trap: %5s, PCWD: 0x%016x, RD: %02d, RWD: 0x%016x, MA: 0x%016x, MWD: 0x%016x, MWM: 0b%08b, I: 0x%016x (%s)"
-                  (show $ rvfi_trap tok /= 0) -- Trap
-                  (rvfi_pc_wdata tok)    -- PCWD
-                  (rvfi_rd_addr tok)     -- RD
-                  (rvfi_rd_wdata tok)    -- RWD
-                  (rvfi_mem_addr tok)    -- MA
-                  (rvfi_mem_wdata tok)   -- MWD
-                  (rvfi_mem_wmask tok)   -- MWM
-                  (rvfi_insn tok) (pretty (toInteger (rvfi_insn tok))) -- Inst
+    | rvfiIsHalt tok = printf "halt token v%d" (rvfiHaltVersion tok)
```

Comment:
Is the version always sent in a halt packet or just during version negotiation? If the latter, it'd be nice to split this out into v1/legacy/no version and non-v1, with the former not printing a version number.

Reply:
Right now sail+qemu only send it during negotiation, but that can be changed easily.